File: Harmonic_oscillator/Python_codes/Harmonic_oscillator_MC_ground_state_wave_function.py
#%%
import numpy as np
import matplotlib.pyplot as plt
import time
import random
# %%

#------------------------------------------------------------------
# The following equations are written in natural units 
# hcross = c = 1

# Action
def V_Harm(lamda,x):
    return 0.5* m* (omega**2) * (x**2)


def Action(V,x1,x2,x):
    S = 0.5 * m * ( (x1 - x)**2 + (x2 - x)**2) + V(lamda,x)  
    return S  

def Action_total(V,path):
    S = 0
    for i in range(Ntau):
        S += 0.5*m*( (path[(i + 1)%Ntau] - path[i] )**2 ) + V(lamda,path[i])
    return S


def MC_sweep(path0,Ntau,h,V):
    accept_rate = 0
    path = path0

    index = np.arange(0,Ntau,1)
    random.shuffle(index)
    
    for i in range(Ntau):
        t = index[i]
        tmin = (t + Ntau - 1)% Ntau  # periodic boundary conditions
        tplu = (t + 1) % Ntau 
        x_new = path[t] + h * (random.random() - 0.5)
        S_old = Action(V,path[tplu],path[tmin],path[t])
        S_new = Action(V,path[tplu],path[tmin],x_new)
        if random.random() < np.exp( - (S_new - S_old) ):
            path[t] = x_new
            accept_rate += 1/Ntau

    h = h * accept_rate/idrate

    return path, h


# %%

dt = 1
Ntau = int(120/dt) # Number of time slices
lamda = 2
path = []
h = 0.1
m = 1 * dt
omega = 1 * dt 
idrate = 0.8
time_arr = np.arange(0,Ntau,1)
#path0 = np.array([random.random() for i in range(Ntau)])

path0 = np.zeros(Ntau)
path_arr = []
path = path0
path_arr.append(list(path))

for i in range(100):
    path, h = MC_sweep(path,Ntau,h,V_Harm)
    path_arr.append(list(path))

thermalized_path = path
#%%

N = 1000
Nsep = 12
path_arr = []


for i in range(N):
    h = 0.1
    path = thermalized_path
    for j in range(Nsep):
        path, h = MC_sweep(path,Ntau,h,V_Harm)
    path_arr.append(list(path))


path_arr_hist = np.array(path_arr).flatten()
#%%

# Analytical solution
x = np.arange(-3,3,0.01)
psi0 = np.exp(- m*omega * x**2) * np.sqrt(m*omega/np.pi)

#%%
plt.hist(path_arr_hist,bins=500,density=True,histtype="step")
plt.plot(x,psi0)
plt.xlabel("x")
plt.ylabel(r"$|\psi_0|^2$")
plt.savefig("Harmonic_oscillator_ground_state_1000_runs.png",dpi=400)
plt.show()
plt.close()
#%%

plt.plot(path_arr[N-1], time_arr)
plt.xlabel("x")
plt.xlim((-4,4))
plt.ylabel("Time")
plt.savefig("Harmonic_oscillator_single_path.png",dpi=400)
plt.show()
plt.close()
# %%

File: Harmonic_oscillator/Julia_codes/Harmonic_oscillator_MC_ground_state_wave_function.jl
#%%
using Plots
pyplot()
using Random
using LaTeXStrings
# %%

#------------------------------------------------------------------
# The following equations are written in natural units 
# hcross = c = 1

# Action
function Action(x1,x2,x) 
    S = 0.5 * m * ( (x1 - x)^2 + (x2 - x)^2 + (omega^2) * (x^2) ) 
    return S
end  

function Action_total(path)
    S = 0
    for i in 1:Ntau
        S += 0.5*m*( (path[mod1(i + 1,Ntau)] - path[i] )^2 + (omega^2) * (path[i]^2) )
    return S
    end
end

function MC_sweep(path0,Ntau,h)
    #path_arr = []
    accept_rate = 0
    path = path0
    #append!(path_arr,path)

    index = collect(1:Ntau)
    shuffle!(index)
    
    for i in 1:Ntau
        t = index[i]
        tmin = mod1(t + Ntau - 1, Ntau)   # periodic boundary conditions
        tplu = t % Ntau + 1 
        x_new = path[t] + h * (rand() - 0.5)
        S_old = Action(path[tplu],path[tmin],path[t])
        S_new = Action(path[tplu],path[tmin],x_new)
        if rand() < exp( - (S_new - S_old) )
            path[t] = x_new
            accept_rate += 1/Ntau
            #append!(path_arr,path)
        end
    end
    
    h = h * accept_rate/idrate

    return path, h
end
    


# %%

Ntau = 120 # Number of time slices
path = []
h = 0.1
m = 1
omega = 1 
idrate = 0.8
dt = 1
time_arr = 1:Ntau
#path0 = [rand() for i in 1:Ntau]

path0 = zeros(Ntau)
path_arr = []
path = path0
append!(path_arr,path)

for i in 1:100
    path, h = MC_sweep(path,Ntau,h);
    #append!(path_arr,path)
end

thermalized_path = path
#%%

N = 10000
Nsep = 12
path_arr = []

lock = Threads.SpinLock()
Threads.@threads for i in 1:N
    h = 0.1
    path = thermalized_path
    for j in 1:Nsep
        path, h = MC_sweep(path,Ntau,h)
    end
    Threads.lock(lock)
    push!(path_arr,copy(path))
    Threads.unlock(lock)
end

path_arr_hist = collect(Iterators.flatten(path_arr))
histogram(path_arr_hist)
#%%

# Analytical solution
x = -3:0.01:3
psi0 = exp.(-x.^2)/sqrt(pi)

#%%
histogram(path_arr_hist,bins=1000,label = "MC",normed=true,grid=false)
plot!(x,psi0,label="Analytical",grid=false)
xlabel!("x")
ylabel!(L"$|\psi_0|^2$")
savefig("Harmonic_oscillator_wavefunction.svg")
#%%

#=  
S_arr = []
for i in path_arr
    append!(S_arr,Action_total(i))
end

plot(1:length(path_arr),S_arr)
# %%

#for i in path_arr
#    plt.plot(time_arr,i)
#end
# %%
#plot( path_arr[1], time_arr)
#plot!(path_arr[N-10], time_arr)
plot(path_arr[N], time_arr,grid = false, legend = false)
xlabel!("x")
xlims!((-4,4))
ylabel!("Time")
savefig("Harmonic_oscillator_single_path.svg") =#
# %%
